file/unixfile.go

```diff
@@ -170,63 +176,97 @@ func NewUnixfsFile(ctx context.Context, dserv ipld.DAGService, nd ipld.Node, met
 		return nil, errors.New("unknown node type")
 	}
 
-	// Skip token metadata if the given 'metadata' is false. I.e.,
-	// check if the given 'p' represents a dummy root of a DAG with token metadata.
-	// If yes, get the root node of the user data that is the second child
-	// of the the dummy root. Then set this child node to 'nd'.
-	// Note that a new UnixFS type to indicate existence of metadata will be faster but
-	// a new type causes many changes.
-	if !meta && !rawNode {
-		newNode, err := skipMetadataIfExists(ctx, nd, dserv)
+	var dr uio.DagReader
+	// Keep 'nd' if raw node
+	if !rawNode {
+		// Split metadata node and data node if available
+		dataNode, metaNode, err := checkAndSplitMetadata(ctx, nd, dserv)
 		if err != nil {
 			return nil, err
 		}
-		if newNode == nil {
-			return nil, nil
+
+		// Return just metadata if available
+		if meta {
+			if metaNode == nil {
+				return nil, errors.New("no metadata is available")
+			}
+			nd = metaNode
+		} else {
+			// Select DagReader based on metadata information
+			if metaNode != nil {
+				mdr, err := uio.NewDagReader(ctx, metaNode, dserv)
+				if err != nil {
+					return nil, err
+				}
+				// Read all metadata
+				buf := make([]byte, mdr.Size())
+				_, err = mdr.CtxReadFull(ctx, buf)
+				if err != nil {
+					return nil, err
+				}
+				var rsMeta chunker.RsMetaMap
+				err = json.Unmarshal(buf, &rsMeta)
+				if err != nil {
+					return nil, err
+				}
+				if rsMeta.NumData > 0 && rsMeta.NumParity > 0 && rsMeta.FileSize > 0 {
```

Comment:
I think there could be a better way to know this DAG is reed-solomon based. 
But I think it is ok as of now.